lib/spack/spack/solver/asp.py

```diff
@@ -3367,15 +3369,19 @@ def virtual_on_edge(self, parent_node, provider_node, virtual):
         dependencies[0].update_virtuals((virtual,))
 
     def reorder_flags(self):
-        """Order compiler flags on specs in predefined order.
-
-        We order flags so that any node's flags will take priority over
-        those of its dependents.  That is, the deepest node in the DAG's
-        flags will appear last on the compile line, in the order they
-        were specified.
+        """For each spec, determine the order of compiler flags applied to it.
 
         The solver determines which flags are on nodes; this routine
-        imposes order afterwards.
+        imposes order afterwards. The order is:
```

Comment:
One concern I have with this is that we aren't accounting for the override precedence that was added in the `cc` wrapper in https://github.com/spack/spack/pull/37376.  I *think* what we will probably want at some point is not a strict first/last rule, but a "this flag overrides that one" rule.

The position to put a flag in to force it to override depends on the flag type, and:
1. you may have multiple types of flags (e.g. both `-l` and `-L`/`-I` flags) in the same group, and
2. you can't split a flag group apart b/c you want to preserve internal ordering

I think we'd need to do some surgery on flag groups to do this right, but that is beyond the scope here. I *do* think we should at least preserve DAG precedence (i.e. children come after parents on the CLI) -- see below -- since that  behavior was present before.

Reply:
Presumably the severable part of this you want done here is addressed by https://github.com/spack/spack/pull/41049#discussion_r1593099877